packages/flutter_test/lib/src/controller.dart

```diff
@@ -779,41 +810,142 @@ abstract class WidgetController {
+    }
+    if (elements.length > 1) {
+      throw FlutterError('The finder "$finder" (used in a call to "$callee()") ambiguously found multiple matching widgets. The "$callee()" method needs a single target.');
+    }
+    final Element element = elements.single;
+    final RenderObject? renderObject = element.renderObject;
+    if (renderObject == null) {
+      throw FlutterError(
+        'The finder "$finder" (used in a call to "$callee()") found an element, but it does not have a corresponding render object. '
+        'Maybe the element has not yet been rendered?'
+      );
+    }
+    if (renderObject is! RenderBox) {
+      throw FlutterError(
+        'The finder "$finder" (used in a call to "$callee()") found an element whose corresponding render object is not a RenderBox (it is a ${renderObject.runtimeType}: "$renderObject"). '
+        'Unfortunately "$callee()" only supports targetting widgets that correspond to RenderBox objects in the rendering.'
+      );
+    }
     final RenderBox box = element.renderObject! as RenderBox;
-    return box.localToGlobal(sizeToPoint(box.size));
+    final Offset location = box.localToGlobal(sizeToPoint(box.size));
+    if (warnIfMissed) {
+      final HitTestResult result = HitTestResult();
+      binding.hitTest(result, location);
+      bool found = false;
+      for (final HitTestEntry entry in result.path) {
+        if (entry.target == box) {
+          found = true;
+          break;
+        }
+      }
+      if (!found) {
+        bool outOfBounds = false;
+        if (binding.renderView != null && binding.renderView.size != null) {
+          outOfBounds = !(Offset.zero & binding.renderView.size).contains(location);
+        }
+        if (hitTestWarningShouldBeFatal) {
+          throw FlutterError.fromParts(<DiagnosticsNode>[
+            ErrorSummary('Finder specifies a widget that would not receive pointer events.'),
+            ErrorDescription('A call to $callee() with finder "$finder" derived an Offset ($location) that would not hit test on the specified widget.'),
+            ErrorHint('Maybe the widget is actually off-screen, or another widget is obscuring it, or the widget cannot receive pointer events.'),
+            if (outOfBounds)
+              ErrorHint('Indeed, $location is outside the bounds of the root of the render tree, ${binding.renderView.size}.'),
+            box.toDiagnosticsNode(name: 'The finder corresponds to this RenderBox', style: DiagnosticsTreeStyle.singleLine),
+            ErrorDescription('The hit test result at that offset is: $result'),
+            ErrorDescription('If you expected this target not to be able to receive pointer events, pass "warnIfMissed: false" to "$callee()".'),
+            ErrorDescription('To make this error into a non-fatal warning, set WidgetController.hitTestWarningShouldBeFatal to false.'),
+          ]);
+        }
+        printToConsole(
```

Comment:
Isn't this going to be spammy for existing apps? Why not just fail if the framework was configured to fail on this condition and otherwise be silent? (Would be cool if we could opt new apps/packages into the fatal mode, though...)

Reply:
Yes, it will be spammy. I think that being silent by default makes this largely worthless. I considered making it fatal by default but that's too much, I think. The warning is actionable, so it doesn't violate our guidelines. I think it's a reasonable compromise. In practice, it does find bugs (at least if my experience with our own tests is any indication), so I think it's useful to print for existing tests.